_posts/2022-12-12-deltalake-spark-testing-learnings.md

```diff
@@ -0,0 +1,232 @@
+
+If you are using a tool like Snowflake, you can easily write integration tests for things like your procedures or UDFs if you use them to run your data transformation tasks. With Databricks however, the landscape is a bit different in some aspects in that we can also dip our hands into unit testing, so lets find out why!
+
+Before we commence, please note that any examples and discussions below regarding unit testing will be focused on users who are currently writing their code in `python 3` and running orchestrations via `Databricks Workflows` as `Delta Live Tables` are a separate topic altogether. 
+
+The following topics will be covered here as we make our way to unit testing. If you are already familiar with points 1-3, them skip to the last section at the bottom `Unit Testing`: 
+
+1. Delta Lake
+2. Apache Spark & PySpark
+3. Integration Testing 
+4. Unit Testing
+
+
+### Delta Lake
+
+
+<img height="100px" src="/img/blog/databricks/delta-lake.png" />
+<br/><br/>
+
+Delta Lake is the open source storage framework providing key features such as ACID transactions and time travel features. Delta Lake is not only compatible with Spark but is also what Databricks utilises for data storage and table management. As a result the types of tables we will be dealing with for the most part are delta tables. The storage format of these delta tables are basically parquet files. 
+
+You can see more details in the official docs [delta.io](https://delta.io/) if you wish to learn more. Interesting stuff!
+
+
+### Apache Spark
+
+<img height="100px" src="/img/blog/databricks/apache-spark.png" />
+<br/><br/>
+
+Apache Spark is an analytics query engine which is good for large scale and batched data workloads, machine learning and real time analytics. Long story short, Databricks is built on Spark and PySpark and hence knowledge in those areas could help quite significantly. 
+
+Apache Spark also supports APIs for developing in R, Java, Scala and Python. This translates to Databricks as well as you find that our notebooks can be written purely as a bunch of SQL commands or in one of those languages as listed above but alas, we will be focusing on Python and PySpark (its just the python library that interfaces Apache Spark).
+
+
+### Integration Testing
+
+While integration testing on Databricks is not covered in this blog, is essential in my opnion that time and effort are still dedicated into comprehensively writing these out as well using a library like `Great Expectations`, running in a notebook and serving as an additional step at the gold layer of your workflow. This serves its purpose to verify that data quality standards are constantly met and if anything unexpected comes up, we fail the workflow and generate a detailed visual report for users to view and understand what the unexpected data is before it gets to the consumers. 
```

Comment:
```suggestion
While integration testing on Databricks is not covered in this blog, is essential in my opnion that time and effort are still dedicated into comprehensively writing these out as well using a library like [`Great Expectations`](url_here), running in a notebook and serving as an additional step at the gold layer of your workflow. This serves its purpose to verify that data quality standards are constantly met and if anything unexpected comes up, we fail the workflow and generate a detailed visual report for users to view and understand what the unexpected data is before it gets to the consumers. 
```

Suggestion - add a link to Great Expectations.

Reply:
added